diff --git a/gson/src/test/java/com/google/gson/stream/JsonWriterTest.java b/gson/src/test/java/com/google/gson/stream/JsonWriterTest.java
index 13857696..d9435f56 100644
--- a/gson/src/test/java/com/google/gson/stream/JsonWriterTest.java
+++ b/gson/src/test/java/com/google/gson/stream/JsonWriterTest.java
@@ -970,7 +970,9 @@ public final class JsonWriterTest {
     writer.beginArray();
     writer.endArray();
     writer.close();
+    assertThat(stringWriter.toString()).isEqualTo("[]");
     writer.close();
+    assertThat(stringWriter.toString()).isEqualTo("[]");
   }
 
   @Test
