diff --git a/gson/src/test/java/com/google/gson/MixedStreamTest.java b/gson/src/test/java/com/google/gson/MixedStreamTest.java
index 42fed496..23c887d1 100644
--- a/gson/src/test/java/com/google/gson/MixedStreamTest.java
+++ b/gson/src/test/java/com/google/gson/MixedStreamTest.java
@@ -197,7 +197,7 @@ public final class MixedStreamTest {
   }
 
   @Test
-  public void testWriteHtmlSafe() {
+  public void testWriteHtmlSafeWithEscaping() {
     List<String> contents = Arrays.asList("<", ">", "&", "=", "'");
     Type type = new TypeToken<List<String>>() {}.getType();
 
@@ -205,8 +205,14 @@ public final class MixedStreamTest {
     new Gson().toJson(contents, type, new JsonWriter(writer));
     assertThat(writer.toString())
         .isEqualTo("[\"\\u003c\",\"\\u003e\",\"\\u0026\",\"\\u003d\",\"\\u0027\"]");
+  }
+
+  @Test
+  public void testWriteHtmlSafeWithoutEscaping() {
+    List<String> contents = Arrays.asList("<", ">", "&", "=", "'");
+    Type type = new TypeToken<List<String>>() {}.getType();
 
-    writer = new StringWriter();
+    StringWriter writer = new StringWriter();
     new GsonBuilder().disableHtmlEscaping().create().toJson(contents, type, new JsonWriter(writer));
     assertThat(writer.toString()).isEqualTo("[\"<\",\">\",\"&\",\"=\",\"'\"]");
   }
