diff --git a/store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java b/store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java
index 59eb14033..c4a9aeb81 100644
--- a/store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java
+++ b/store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java
@@ -265,7 +265,7 @@ public class AutoSwitchHAService extends DefaultHAService {
     }
 
     @Override
-    public int inSyncReplicasNums(final long masterPutWhere) {
+    public synchronized int inSyncReplicasNums(final long masterPutWhere) {
         return syncStateSet.size();
     }
 
