diff --git a/client/src/main/java/org/apache/rocketmq/client/common/ThreadLocalIndex.java b/client/src/main/java/org/apache/rocketmq/client/common/ThreadLocalIndex.java
index 3a086c13d..c15cdbfad 100644
--- a/client/src/main/java/org/apache/rocketmq/client/common/ThreadLocalIndex.java
+++ b/client/src/main/java/org/apache/rocketmq/client/common/ThreadLocalIndex.java
@@ -35,9 +35,6 @@ public class ThreadLocalIndex {
 
     public void reset() {
         int index = Math.abs(random.nextInt(Integer.MAX_VALUE));
-        if (index < 0) {
-            index = 0;
-        }
         this.threadLocalIndex.set(index);
     }
 
