diff --git a/tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java b/tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java
index 4b354ac7f..02158ae54 100644
--- a/tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java
+++ b/tools/src/main/java/org/apache/rocketmq/tools/command/controller/CleanControllerBrokerDataSubCommand.java
@@ -87,7 +87,7 @@ public class CleanControllerBrokerDataSubCommand implements SubCommand {
         }
 
         if (!isCleanLivingBroker && StringUtils.isEmpty(clusterName)) {
-            throw new IllegalArgumentException("cleanLivingBroker option is false,clusterName option can not be empty.");
+            throw new IllegalArgumentException("cleanLivingBroker option is false, clusterName option can not be empty.");
         }
 
         try {
