diff --git a/assertj-core/src/main/java/org/assertj/core/internal/StandardComparisonStrategy.java b/assertj-core/src/main/java/org/assertj/core/internal/StandardComparisonStrategy.java
index 84f29f11e..4107c62d1 100644
--- a/assertj-core/src/main/java/org/assertj/core/internal/StandardComparisonStrategy.java
+++ b/assertj-core/src/main/java/org/assertj/core/internal/StandardComparisonStrategy.java
@@ -15,6 +15,7 @@ package org.assertj.core.internal;
 import static org.assertj.core.util.Preconditions.checkArgument;
 
 import java.util.Collection;
+import java.util.Date;
 import java.util.Iterator;
 import java.util.Set;
 import java.util.TreeSet;
@@ -105,6 +106,10 @@ public class StandardComparisonStrategy extends AbstractComparisonStrategy {
           return java.util.Arrays.deepEquals((Object[]) actual, (Object[]) other);
       }
     }
+
+    if (actual instanceof Date && other instanceof Date) {
+      return ((Date) actual).compareTo((Date) other) == 0;
+    }
     return actual.equals(other);
   }
 
