diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/PropertyCacheFile.java b/src/main/java/com/puppycrawl/tools/checkstyle/PropertyCacheFile.java
index 55f518207..c53bc6d5c 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/PropertyCacheFile.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/PropertyCacheFile.java
@@ -143,8 +143,21 @@ public final class PropertyCacheFile {
      */
     public void persist() throws IOException {
         final Path path = Paths.get(fileName);
-        final Path directory = path.getParent();
+        Path directory = path.getParent();
+
         if (directory != null) {
+            if (Files.isSymbolicLink(directory)) {
+                final Path actualDir = directory.toRealPath();
+
+                if (Files.isDirectory(actualDir)) {
+                    directory = actualDir;
+                }
+                else {
+                    throw new IOException(
+                            "Resolved symbolic link " + directory
+                                    + " is not a directory.");
+                }
+            }
             Files.createDirectories(directory);
         }
         try (OutputStream out = Files.newOutputStream(path)) {
