diff --git a/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/StdTypeResolverBuilder.java b/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/StdTypeResolverBuilder.java
index 32da2e179..cfc8fb058 100644
--- a/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/StdTypeResolverBuilder.java
+++ b/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/StdTypeResolverBuilder.java
@@ -37,6 +37,11 @@ public class StdTypeResolverBuilder
     protected boolean _typeIdVisible = false;
 
     /**
+     * 
+     * Boolean value configured through {@link JsonTypeInfo#requireTypeIdForSubtypes}.
+     * If this value is not {@code null}, this value should override the global configuration of
+     * {@link com.fasterxml.jackson.databind.MapperFeature#REQUIRE_TYPE_ID_FOR_SUBTYPES}. 
+     * 
      * @since 2.16 (backported from Jackson 3.0)
      */
     protected Boolean _requireTypeIdForSubtypes;
@@ -466,7 +471,10 @@ public class StdTypeResolverBuilder
 
     /**
      * Determines whether strict type ID handling should be used for this type or not.
-     * This will be enabled when either the type has type resolver annotations or if
+     * This will be enabld as configured by {@link JsonTypeInfo#requireTypeIdForSubtypes()}
+     * unless its value is {@link com.fasterxml.jackson.annotation.OptBoolean#DEFAULT}. 
+     * In case the value of {@link JsonTypeInfo#requireTypeIdForSubtypes()} is {@code OptBoolean.DEFAULT},
+     * this will be enabled when either the type has type resolver annotations or if
      * {@link com.fasterxml.jackson.databind.MapperFeature#REQUIRE_TYPE_ID_FOR_SUBTYPES}
      * is enabled.
      *
@@ -479,6 +487,10 @@ public class StdTypeResolverBuilder
      * @since 2.15
      */
     protected boolean _strictTypeIdHandling(DeserializationConfig config, JavaType baseType) {
+        // [databind#3877]: per-type strict type handling, since 2.16
+        if (_requireTypeIdForSubtypes != null && baseType.isConcrete()) {
+            return _requireTypeIdForSubtypes;
+        }
         if (config.isEnabled(MapperFeature.REQUIRE_TYPE_ID_FOR_SUBTYPES)) {
             return true;
         }
