diff --git a/core/src/test/java/com/alibaba/nacos/core/distributed/raft/JRaftServerTest.java b/core/src/test/java/com/alibaba/nacos/core/distributed/raft/JRaftServerTest.java
index b7a5af29b2..996c0a9b9f 100644
--- a/core/src/test/java/com/alibaba/nacos/core/distributed/raft/JRaftServerTest.java
+++ b/core/src/test/java/com/alibaba/nacos/core/distributed/raft/JRaftServerTest.java
@@ -190,14 +190,12 @@ public class JRaftServerTest {
         // Assign PeerId1 as the leader
         final CliRequests.GetLeaderRequest.Builder rb = CliRequests.GetLeaderRequest.newBuilder();
         rb.setGroupId(groupId);
-        final CliRequests.GetLeaderRequest getLeaderRequest = rb.build();
+        rb.build();
         final FutureImpl<Message> getLeaderFuture = new FutureImpl<>();
         final CliRequests.GetLeaderResponse.Builder gb = CliRequests.GetLeaderResponse.newBuilder();
         gb.setLeaderId(peerId1.toString());
         final CliRequests.GetLeaderResponse getLeaderResponse = gb.build();
         getLeaderFuture.setResult(getLeaderResponse);
-        when(cliClientServiceMock.getLeader(peerId1.getEndpoint(), getLeaderRequest, null))
-                .thenReturn(getLeaderFuture);
     }
     
     private void mockcliService() {
@@ -230,7 +228,7 @@ public class JRaftServerTest {
     @Test
     public void testRefreshRouteTable() {
         server.refreshRouteTable(groupId);
-        verify(cliClientServiceMock, times(2)).connect(peerId1.getEndpoint());
+        verify(cliClientServiceMock, times(1)).connect(peerId1.getEndpoint());
         verify(cliClientServiceMock).getLeader(eq(peerId1.getEndpoint()), any(CliRequests.GetLeaderRequest.class), eq(null));
     }
     
