diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheckTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheckTest.java
index c300bbb5a..3ea881caa 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheckTest.java
@@ -467,4 +467,22 @@ public class AbbreviationAsWordInNameCheckTest extends AbstractModuleTestSupport
                 expected);
     }
 
+    @Test
+    public void testInputAbbreviationAsWordInNameTypeSnakeStyle() throws Exception {
+        final String[] expected = {
+            "13:20: " + getWarningMessage("FLAG_IS_FIRST_RUN", 4),
+            "16:17: " + getWarningMessage("HYBRID_LOCK_PATH", 4),
+            "21:17: " + getWarningMessage("__DEMOS__TESTS_VAR", 4),
+            "28:16: " + getWarningMessage("TESTING_FAM_23456", 4),
+            "33:16: " + getWarningMessage("TESTING_23456_FAM", 4),
+            "38:16: " + getWarningMessage("_234VIOLATION", 4),
+            "41:16: " + getWarningMessage("VIOLATION23456", 4),
+            "72:21: " + getWarningMessage("getIsFIRST_Run", 4),
+            "77:21: " + getWarningMessage("getBoolean_VALUES", 4),
+        };
+
+        verifyWithInlineConfigParser(
+                getPath("InputAbbreviationAsWordInNameTypeSnakeStyle.java"), expected);
+    }
+
 }
